File: trees/kthSmallestElementInABST.java
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/

package trees;

public class kthSmallestElementInABST {
	public static void main(String[] args) {

		System.out.println("This is an Interactive problem please Test the Solution In Leetcode");

	}
}

class SolutionKSEIABST {

	// Submitted by @Jiganesh
	// TC : O(N)
	// SC : O(1)
	public int kthSmallest(TreeNode root, int k) {

		int[] array = new int[2];
		array[0] = k;
		traverse(array, root);
		return array[1];
	}

	public int[] traverse(int[] array, TreeNode root) {
		if (root != null) {
			traverse(array, root.left);
			if (--array[0] == 0)
				array[1] = root.val;
			traverse(array, root.right);
		}
		return array;
	}
}

File: trees/kthSmallestElementInABST.py
# https://leetcode.com/problems/kth-smallest-element-in-a-bst/

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    # Runtime: 48 ms, faster than 91.89% of Python3 online submissions for Kth Smallest Element in a BST.
    # Memory Usage: 18.2 MB, less than 13.38% of Python3 online submissions for Kth Smallest Element in a BST.
    def kthSmallest(self, root, k: int) -> int:
        self.k = k
        self.kthElement = -1

        def helper(root):
            if root:
                helper(root.left)
                self.k -= 1
                if self.k == 0:
                    self.kthElement = root.val
                    return self.kthElement
                helper(root.right)

        helper(root)
        return self.kthElement
